pkg/validator/validators.go

```diff
@@ -244,25 +254,149 @@ func (kc *policyValidator) upstreamTrafficSettingValidator(req *admissionv1.Admi
 		// duplicate detected
-		return nil, fmt.Errorf("UpstreamTrafficSetting %s/%s conflicts with %s/%s since they have the same host %s", ns, upstreamTrafficSetting.ObjectMeta.GetName(), ns, matchingUpstreamTrafficSetting.ObjectMeta.GetName(), matchingUpstreamTrafficSetting.Spec.Host)
+		return nil, fmt.Errorf("upstreamTrafficSetting %s/%s conflicts with %s/%s since they have the same host %s", ns, upstreamTrafficSetting.ObjectMeta.GetName(), ns, matchingUpstreamTrafficSetting.ObjectMeta.GetName(), matchingUpstreamTrafficSetting.Spec.Host)
 	}
 
 	// Validate rate limiting config
 	rl := upstreamTrafficSetting.Spec.RateLimit
 	if rl != nil && rl.Local != nil && rl.Local.HTTP != nil {
 		if _, ok := xds_type.StatusCode_name[int32(rl.Local.HTTP.ResponseStatusCode)]; !ok {
-			return nil, fmt.Errorf("Invalid responseStatusCode %d. See https://www.envoyproxy.io/docs/envoy/latest/api-v3/type/v3/http_status.proto#enum-type-v3-statuscode for allowed values",
+			return nil, fmt.Errorf("invalid responseStatusCode %d. See https://www.envoyproxy.io/docs/envoy/latest/api-v3/type/v3/http_status.proto#enum-type-v3-statuscode for allowed values",
 				rl.Local.HTTP.ResponseStatusCode)
 		}
 	}
 	for _, route := range upstreamTrafficSetting.Spec.HTTPRoutes {
 		if route.RateLimit != nil && route.RateLimit.Local != nil {
 			if _, ok := xds_type.StatusCode_name[int32(route.RateLimit.Local.ResponseStatusCode)]; !ok {
-				return nil, fmt.Errorf("Invalid responseStatusCode %d. See https://www.envoyproxy.io/docs/envoy/latest/api-v3/type/v3/http_status.proto#enum-type-v3-statuscode for allowed values",
+				return nil, fmt.Errorf("invalid responseStatusCode %d. See https://www.envoyproxy.io/docs/envoy/latest/api-v3/type/v3/http_status.proto#enum-type-v3-statuscode for allowed values",
 					route.RateLimit.Local.ResponseStatusCode)
 			}
 		}
 	}
 
 	return nil, nil
 }
+
+var validMRCTransitions = map[string][]string{
+	constants.MRCStateValidatingRollout:  {constants.MRCStateIssuingRollout, constants.MRCStateError},
+	constants.MRCStateIssuingRollout:     {constants.MRCStateActive, constants.MRCStateError},
+	constants.MRCStateActive:             {constants.MRCStateValidatingRollback, constants.MRCStateError},
+	constants.MRCStateValidatingRollback: {constants.MRCStateIssuingRollback, constants.MRCStateError},
+	constants.MRCStateIssuingRollback:    {constants.MRCStateInactive, constants.MRCStateError},
+	constants.MRCStateInactive:           {},
+	constants.MRCStateError:              {},
+	"":                                   {constants.MRCStateValidatingRollout, constants.MRCStateError},
+}
+
+// meshRootCertificateValidator validates the MeshRootCertificate custom resource
+func (cv *configValidator) meshRootCertificateValidator(req *admissionv1.AdmissionRequest) (*admissionv1.AdmissionResponse, error) {
+	log.Debug().Msg("made it to MRC validator")
+	switch req.Operation {
```

Comment:
we can validate there are no deletions if it is in some form of an active state.

Reply:
Would this prevent deletion on uninstall?

Reply:
Not sure how that would work actually since helm's ordering would delete the service and pods first anyways

https://github.com/helm/helm/blob/12f1bc0acdeb675a8c50a78462ed3917fb7b2e37/pkg/releaseutil/kind_sorter.go#L72.

Might be worth testing out. If so, maybe we add a custom uninstall portion that removes the webhooks prior to the helm uninstall

Reply:
Will explore and track this in a separate PR/issue